File: Java/Quicksort.java
import java.util.Arrays;

public class Quicksort {

    public static void quicksort(int[] arr, int inicio, int fim) {
        if (inicio < fim) {
            int p = particionar(arr, inicio, fim);
            quicksort(arr, inicio, p - 1);
            quicksort(arr, p + 1, fim);
        }
    }

    public static int particionar(int[] arr, int inicio, int fim) {
        int pivo = arr[fim];
        int i = inicio - 1;

        for (int j = inicio; j < fim; j++) {
            if (arr[j] <= pivo) {
                i++;
                trocar(arr, i, j);
            }
        }
        trocar(arr, i + 1, fim);
        return i + 1;
    }

    public static void trocar(int[] arr, int x, int y) {
        int temp = arr[x];
        arr[x] = arr[y];
        arr[y] = temp;
    }

    public static void main(String[] args) {
        int[] lista = {5, 3, 8, 1, 9, 2, 7};

        System.out.println("--- Ordenacao com Quicksort ---");
        System.out.println("Antes:  " + Arrays.toString(lista));
        quicksort(lista, 0, lista.length - 1);
        System.out.println("Depois: " + Arrays.toString(lista));
    }
}

File: JavaScript/quicksort.js
function Quicksort(arr, inicio = 0, fim = arr.length - 1) {
  if (inicio < fim) {
    const p = Particionar(arr, inicio, fim);
    Quicksort(arr, inicio, p - 1);
    Quicksort(arr, p + 1, fim);
  }
  return arr;
}

function Particionar(arr, inicio, fim) {
  const pivo = arr[fim];
  let i = inicio - 1;

  for (let j = inicio; j < fim; j++) {
    if (arr[j] <= pivo) {
      i++;
      [arr[i], arr[j]] = [arr[j], arr[i]];
    }
  }
  [arr[i + 1], arr[fim]] = [arr[fim], arr[i + 1]];
  return i + 1;
}

const lista = [5, 3, 8, 1, 9, 2, 7];

console.log("--- Ordenação com Quicksort ---");
console.log("Antes:  " + lista.join(", "));
Quicksort(lista);
console.log("Depois: " + lista.join(", "));